wallet-new/src/Cardano/Wallet/Kernel/CoinSelection/Generic/Fees.hs

```diff
@@ -35,107 +36,167 @@ data FeeOptions dom = FeeOptions {
+    :: forall dom. CoinSelDom dom
+    => FeeOptions dom
+    -> [UtxoEntry dom]
+    -> [Output dom]
+    -> [Value dom]
+    -> Except CoinSelHardErr ([UtxoEntry dom], [Output dom], [Value dom])
+receiverPaysFee feeOptions inps outs chgs = do
+    let totalFee = feeUpperBound feeOptions inps outs chgs
+    outs' <- mapM go . divvyFee outVal totalFee $ outs
+    return (inps, outs', chgs)
   where
-    go :: (Fee dom, CoinSelResult dom)
-       -> Except CoinSelHardErr (CoinSelResult dom)
-    go (fee, cs) =
-        case outSubFee fee (coinSelRequest cs) of
+    go
+        :: (Fee dom, Output dom)
+        -> Except CoinSelHardErr (Output dom)
+    go (fee, out) =
+        case outSubFee fee out of
           Just newOut ->
-            return $ cs { coinSelOutput = newOut }
+            return newOut
           Nothing ->
             throwError $
-              CoinSelHardErrOutputCannotCoverFee (pretty (coinSelRequest cs)) (pretty fee)
+              CoinSelHardErrOutputCannotCoverFee (pretty out) (pretty fee)
 
 {-------------------------------------------------------------------------------
   Sender pays fee
 -------------------------------------------------------------------------------}
 
-senderPaysFee :: (Monad m, CoinSelDom (Dom utxo))
-              => (Value (Dom utxo) ->
-                   CoinSelT utxo CoinSelHardErr m (Maybe (UtxoEntry (Dom utxo))))
-              -> Fee (Dom utxo)
-              -> [CoinSelResult (Dom utxo)]
-              -> CoinSelT utxo CoinSelHardErr m
-                   ([CoinSelResult (Dom utxo)], SelectedUtxo (Dom utxo), Value (Dom utxo))
-senderPaysFee pickUtxo totalFee css = do
-    let (css', remainingFee) = feeFromChange totalFee css
-    (additionalUtxo, additionalChange) <- coverRemainingFee pickUtxo remainingFee
-    return (css', additionalUtxo, additionalChange)
-
-coverRemainingFee :: forall utxo m. (Monad m, CoinSelDom (Dom utxo))
-                  => (Value (Dom utxo) -> CoinSelT utxo CoinSelHardErr m (Maybe (UtxoEntry (Dom utxo))))
-                  -> Fee (Dom utxo)
-                  -> CoinSelT utxo CoinSelHardErr m (SelectedUtxo (Dom utxo), Value (Dom utxo))
+senderPaysFee
+    :: (Monad m, CoinSelDom (Dom utxo))
+    => PickUtxo m utxo
+    -> FeeOptions (Dom utxo)
+    -> [UtxoEntry (Dom utxo)]
+    -> [Output (Dom utxo)]
+    -> [Value (Dom utxo)]
+    -> CoinSelT utxo CoinSelHardErr m ([UtxoEntry (Dom utxo)], [Output (Dom utxo)], [Value (Dom utxo)])
+senderPaysFee pickUtxo feeOptions = go
+  where
+    go inps outs chgs = do
+        -- 1/
+        -- We compute fees using all inputs, outputs and changes since
+        -- all of them have an influence on the fee calculation.
+        let fee = feeUpperBound feeOptions inps outs chgs
+
+        -- 2/
+        -- We try to cover fee with the available change by substracting equally
+        -- across all inputs. There's no fairness in that in the case of a
+        -- multi-account transaction. Everyone pays the same part.
```

Comment:
TODO: Amend this comment. Not really on the spot. We still divvy fee proportionally across all changes. Thus, the output:change ratio is more-or-less preserved. In the end, it may be less accurate but fairness still applies here.